src/App/PropertyLinks.cpp

```diff
@@ -275,12 +334,141 @@ bool PropertyLinkBase::_updateElementReference(DocumentObject *feature,
+        subname = shadow.first.c_str();
+    else if (shadow.second.size())
+        subname = shadow.second.c_str();
+    else
+        subname = sub.c_str();
+    GeoFeature* geo = 0;
+    const char* element = 0;
+    auto ret = GeoFeature::resolveElement(obj,
+                                          subname,
+                                          elementName,
+                                          true,
+                                          GeoFeature::ElementNameType::Export,
+                                          feature,
+                                          &element,
+                                          &geo);
+    if (!ret || !geo || !element || !element[0]) {
+        if (elementName.second.size())
+            shadow.second.swap(elementName.second);
+        return false;
+    }
+
+    if (_ElementRefs.insert(geo).second)
+            _ElementRefMap[geo].insert(this);
+
+        if (!reverse) {
+        if (elementName.first.empty()) {
+            shadow.second.swap(elementName.second);
+            return false;
+        }
+        if (shadow == elementName)
+                return false;
+        }
+
+    bool missing = Data::hasMissingElement(elementName.second.c_str());
+    if (feature == geo && (missing || reverse)) {
+        // If the referenced element is missing, or we are generating element
+        // map for the first time, or we are re-generating the element map due
+        // to version change, i.e. 'reverse', try search by geometry first
+        const char* oldElement = Data::findElementName(shadow.second.c_str());
+        if (!Data::hasMissingElement(oldElement)) {
+//            const auto& names = geo->searchElementCache(oldElement);
+            std::vector<std::string> names; // searchElementCache isn't implemented.
+            if (names.size()) {
+                missing = false;
+                std::string newsub(subname, strlen(subname) - strlen(element));
+                newsub += names.front();
+                GeoFeature::resolveElement(obj,
+                                           newsub.c_str(),
+                                           elementName,
+                                           true,
+                                           GeoFeature::ElementNameType::Export,
+                                           feature);
+                const auto& oldName = shadow.first.size() ? shadow.first : shadow.second;
+                const auto& newName =
+                    elementName.first.size() ? elementName.first : elementName.second;
+                if (oldName != newName) {
+                    FC_WARN(propertyName(this)
+                            << " auto change element reference " << ret->getFullName() << " "
+                            << oldName << " -> " << newName);
+                }
+            }
+        }
+    }
+
+    if (notify)
+            aboutToSetValue();
+
+        auto updateSub = [&](const std::string& newSub) {

```

Comment:
Indentation is wonky here -- coupled with that single-line conditional above, it looks sketchy.